Join/Create Classroom functionality fully implemented w/ Signup/Profile refactor

mykulls opened the pull request:
### Currently takes the part of 3 components:

- Create Classroom, a settings page for the teacher to edit things for the classroom prior to creation
- Join Classroom, which is housed under HomeScreen.js (although it can be moved later)
- Classroom - the main classroom page, housed under ClassroomHome.js

**Create Classroom:**
- create classroom button is currently visible when the user is set to a teacher on the **Home Screen**
- clicking it brings you to a new page (CreateClassroomScreen.js) that gives you settings for the class that you can change
- there is only one input field for the class's name (classTitle in Firebase) currently
- upon clicking create, it'll generate a 6-letter code and a new document will be created in classrooms on Firebase with said code as a custom ID
- adds both the teacher's ID to the classroom's document and the classroom's code to the teacher's document via arrayUnion()

**Join Classroom:**
- join classroom button is currently visible when the user is set to a student on the **Home Screen**, with a TextInput field where you input the 6-letter code
- upon entering a valid code, you're brought to the class's page just like create classroom does
- adds both the student's ID to the classroom's document and the classroom's code to the student's document via arrayUnion(), however currently it does this **every single time** (which isn't a good idea for the future), but it doesn't technically cause issues because arrayUnion() only adds unique items to the Firebase arrays

**ClassroomHome:**
- currently just welcomes the user by taking its name and displaying the classroom's title and code
- takes in the code and classroomInfo (Firebase document fields) via navigation props

**Miscellaneous:**
- the extra commit was b/c i accidentally committed my gitignore and package-lock
- also edited AppNavigation.js to include ClassroomHome and CreateClassroomScreen
- the create/join functionality is housed on the home screen currently, but it is certainly possible to move them in the future
- lots of legacy code from old Classroom.js file still exists, cleanup necessary later

mykulls committed:
Join/Create Classroom functionality created

mykulls committed:
revert gitignore/package lock change

KarenMTan commented on src/screens/Classroom/ClassroomHome.js:
```diff
@@ -33,39 +35,68 @@ const styles = StyleSheet.create({
-}; */
+  // const [showLoading, setShowLoading] = useState(false);
+
+  async function getUserData() {
```
Let's use a promise chain to retrieve data from Firestore: 

useEffect(() => {
Firestore().collection('users')
.doc(uid)
.get()
.then((document) => {
  if(document.exists) {
     return document.data();
}
  return null;
})
.then((data) => {
  setUserState(data);
})
.catch((e) => {
  Alert.alert(e.message);
});
}, []);

_This is the same asynchronous code but written with promise chaining (I wrote this in a text editor, so you may need to make some changes to ensure that it is in fact correct)._

You can learn more about Promise Chaining here: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Using_promises

Promises are another type of asynchronous programming supported by JavaScript. The _return_value_ of each stage is passed into the next `.then((_return_value_))`.

KarenMTan commented on src/screens/Classroom/ClassroomHome.js:
```diff
@@ -33,39 +35,68 @@ const styles = StyleSheet.create({
+    getUserData();
+  }, [userState]);
+
+  if (initializing) return null;
```
When we wrap the Firebase data retrieval within a useEffect(), I believe we no longer need the `initializing` variable, but please double check lol.

KarenMTan commented on src/navigation/AppNavigation.js:
```diff
@@ -10,6 +10,7 @@ import CreateEvaluationScreen from '../screens/CreateEvaluation';
 import GenericFormDemoScreen from '../screens/GenericFormDemo';
 import AnnouncementsScreen from '../screens/Announcements';
 import AssignmentListScreen from '../screens/AssignmentList';
+import { ClassroomHome, CreateClassroomScreen } from '../screens/Classroom';
```
What did you have in mind for the ClassroomHome Screen? I'm thinking in the final implementation, when the user opens the app, the Announcements/Assignments Screen is the "Home Screen", and (for now) the teacher can modify classroom details in Settings or something. The classroom name will be displayed somewhere else (the designers and I will need to work on that because both teachers and students will be able to be part of multiple classrooms - ie. a teacher can teach multiple classes and a student can be part of "Beginning Orchestra" and "Beginning Cello")

mykulls replied:
Sounds good - ClassroomHome was more of a placeholder so I can make sure my backend connections were working. I think most of the code can be moved around when/if necessary.

KarenMTan commented on src/screens/Classroom/CreateClassroomScreen.js:
```diff
@@ -0,0 +1,152 @@
+  const [userState, setUserState] = useState(INITIAL_USER_STATE);
+  const [errorMessage, setErrorMessage] = useState(null);
+
+  async function getUserData() {
```
Our code base uses the promise chaining function style - can you convert this code the same way I did in ClassroomHome.js (although ClassroomHome.js may not be used for now, simply disconnect it from other code rather than deleting it complete for now).

KarenMTan commented on src/screens/Classroom/CreateClassroomScreen.js:
```diff
@@ -0,0 +1,152 @@
+  }, [userState]);
+
+  // creates classroomID
+  function makeid(length) {
```
While it's not likely that this code will match another randomly generated code, I think we should stick to the unique IDs that Firebase generates for us (because those are guaranteed to be unique within the collection).

mykulls replied:
I had it create a unique code so it would be easy to join - I can definitely rework this however necessary.

KarenMTan commented on src/screens/Home/HomeScreen.js:
```diff
@@ -2,15 +2,19 @@
 import Auth from '@react-native-firebase/auth';
 import PropTypes from 'prop-types';
 import UploadFile from '../../components/UploadFile';
+import { INITIAL_USER_STATE } from '../../components';
```
What is this INITIAL_USER_STATE??

mykulls replied:
The initial userData of a user prior to account creation. This might be redundant here but I will double check.

mykulls replied:
Double-checked - it isn't technically necessary, but it is probably a good precaution.

KarenMTan commented on src/screens/Home/HomeScreen.js:
```diff
@@ -23,11 +27,54 @@ const styles = StyleSheet.create({
   function onAuthStateChanged(authUser) {
     setUser(authUser);
-    if (initializing) setInitializing(false);
+  }
```
I think the code that you added to HomeScreen.js would be better suited somewhere else (I will have to think about it and talk to designers). Creating and joining a classroom would definitely need to happen during sign up process, so you could work on that for your next sprint.

mykulls replied:
Sounds good - I left it there at the moment because I figured it would be useful to build and easy to test on the Home Screen.

KarenMTan reviewed (commented):
Hi Michael, I included a promise chaining template that you can continue to use when you want to retrieve information from Firebase. 

I need to think about where to include the create/join classroom code and discuss it with the designers, so let's hold on that for now. Working on adding the create and join classroom functionality for teachers and students when they first sign up for their accounts can be part of your next sprint.

Moreover, next time please include screen shots of the simulation screens in your code review. Thanks!!

mykulls committed:
create classroom and signup/profile refactor

mykulls committed:
uninstall react-native-select-multiple

mykulls committed:
remove errorMessage from CreateClassroom

mykulls commented:
# Join/Create Classroom fully implemented, Signup/Profile refactor

**Changes/Fixes from last request:**
- Promise chaining for fetching user data on **HomeScreen.js**, **ClassroomHome.js** (also removed initializing), **CreateClassroomScreen.js**
- Unique code generation (more on this later)

**Important Notes:**
- committed my **package-lock.json**/**package.json** so that all my added dependencies are there - `npm install` to install them
- first extra commit was to remove an extra dependency I had forgotten to uninstall
- second extra commit was for removing `errorMessage` from and adding a comment to **CreateClassroomScreen**
- the create/join functionality is _still_ housed on **HomeScreen** currently, but it is certainly possible to move them in the future
- `CheckBox` import in **CreateClassroomScreen** is "unresolved" according to eslint, but I don't get any errors or warnings when I run the emulator
---
# Documentation

## Create Classroom:
- for `name`, `type`, `term`, `meetDays`, `startDate`, `endDate`, `description`, and `classLength`, there are respective input fields (unstylized). each input field sets variables that change fields for `classroomInfo`, which is then pushed as a document to 'classrooms' in Firestore.
    - `name`, `description`: 
        - `TextInput` that saves in a string state variable (`className` and `classDesc`, respectively)
    - `classLength`: 
        - numeric-only `TextInput` that saves in a string state variable (`length`). a regex is used to check if `length` only contains numbers. 
        - `parseInt()` is used to set `classLength` in classroomInfo, as this is stored as an integer in Firestore.
    - `type`: 
        - `Picker` that saves in a string state variable (`classType`). set to "Group" (`classTypes.default`) by default. 
        - `classTypes` refers to the const enumeration of the possible class types in **const.js**.
    - `term`, `meetDays`: 
        - `CheckBox`es that toggle object state variables (`termToggle`, `dayToggle`) with fields relating to whether a term or day is selected or not (`true`/`false`).
        - when a classroom is created in `create()`, sets of if statements (could be changed to `forEach()` if ordering is not important) push the right terms/days to local arrays (`classTerms`,`days`), which are then added to `classroomInfo`.
    - `startDate`, `endDate`: 
        - `Calendar` inputs that save date strings (format 'yyyy-mm-dd') to string state variables (`classStart`, `classEnd`). 
        - _minor issue:_ `endDate`'s calendar only updates upon changing months, so it looks like you can select dates prior to `startDate` when you actually can't. an if statement ensures any conflicting dates will never lead to a created classroom, but this is messy UI. (see _third_ screenshot below).
- clicking on the "CREATE" button (see _third_ screenshot below) calls the `create()` function 
   - if a classroom can be created, it first generates a `code` with the function `makeid()`, which generates a code and guarantees uniqueness by fetching all documents within 'classrooms' from Firestore, and `forEach()`, if the generated code is not unique, it generates a new code until it is unique.
      - note: `code` is always saved and generated in lowercase
   - if any input fields are not filled in or have errors, an appropriate `Alert` is shown. (see _fourth_ screenshot below).

<img src="https://user-images.githubusercontent.com/69874869/106396276-c7612a00-63bb-11eb-9251-539097038c7b.png"  width="200" /><img src="https://user-images.githubusercontent.com/69874869/106396280-cc25de00-63bb-11eb-8b92-c6123952217f.png"  width="200" /><img src="https://user-images.githubusercontent.com/69874869/106396151-08a50a00-63bb-11eb-926f-da141dcd87e5.png"  width="200" /><img src="https://user-images.githubusercontent.com/69874869/106396843-5cb1ed80-63bf-11eb-80e3-ae2f12fdee3b.png"  width="200" />
- **Miscellaneous:**
    - added comment to note that this screen requires `uid` passed as a `navigation` parameter
    - removed `errorMessage`; instead uses `Alert` for any errors
    - loading this screen might take a while now? consider some type of loading screen/`ActivityIndicator` in the future

## Classroom Home:
- added "COPY CODE" button, which allows you to copy the class code to clipboard
- displays all the class info (unstylized)
- added comment to note that this screen requires `code`, `classroomInfo`, and `uid` passed as `navigation` parameters
<img src="https://user-images.githubusercontent.com/69874869/106396983-0c875b00-63c0-11eb-8b97-ce90d60e4f04.png"  width="200" />

## Signup:
- signup now includes a `Picker` that allows user to choose between Student, Parent, and Teacher (see _first_ screenshot below).
- password is stored in `password` string state variable, which is used to check whether or not the password is valid.
   - 4 regex expressions are used to check numbers, symbols, lowercase, and uppercase letters in the password. if any of these conditions are not met, or if the password is less than 8 characters in length, an appropriate `Alert` is made (see _second_ screenshot below).
- for Student or Parent, a valid code is needed or an appropriate `Alert` is made (see _third_ screenshot below).
    - when `signup()` is called and password is valid, a user is created via `Auth().createUserWithEmailAndPassword()`.
    - if code invalid, show `Alert` and `Auth().currentUser.delete()` ensures that the user isn't saved
    - if Student and code is valid (case insensitive), fetch the classroom doc from Firestore and update the `students` field to include the id of the user
    - then store user in Firestore (no longer updates `classroomIds`, as this has been removed)
    - then sign in and navigate to **ClassroomHome**, passing `code`, the `data` of the fetched classroom as `classroomInfo`, and the `user.uid` (the user's uid) as `uid` as `navigation` parameters
- Teacher does not require a code (see _fourth_ screenshot below)
    - store user in Firestore
    - then sign in and navigate to **HomeScreen**, passing the `user.uid` (the user's uid) as `uid` as a `navigation` parameter

<img src="https://user-images.githubusercontent.com/69874869/106397498-9d5f3600-63c2-11eb-899c-fc21903abb10.png"  width="200" /><img src="https://user-images.githubusercontent.com/69874869/106397509-b667e700-63c2-11eb-970d-e140a8159083.png"  width="200" /><img src="https://user-images.githubusercontent.com/69874869/106397541-edd69380-63c2-11eb-8e11-29efe0727336.png"  width="200" /><img src="https://user-images.githubusercontent.com/69874869/106397558-f8912880-63c2-11eb-9834-b658e3c91935.png"  width="200" />

## Home Screen:
- changed `onAuthStateChanged()` so `initializing` works as intended
- changed user data fetch to Promise chain, but only when user is signed in (via if statement checking if `uid` is set)
- "Join Classroom" button calls `joinClassroom()`
   - checks for valid code, if code invalid, an appropriate `Alert` is shown (just like **SignUpScreen**)
   - otherwise navigate to **ClassroomHome**, passing `code`, fetched classroom info `classroomInfo`, and the user's `uid` as `navigation` parameters 
- "Create Classroom" button navigates to **CreateClassroomScreen**, passing the user's `uid` as a `navigation` parameter
- added comment to note that this screen requires `uid` passed as a `navigation` parameter
- _commented out_: code that forces teacher to create classroom if they haven't created one yet
<img src="https://user-images.githubusercontent.com/69874869/106397193-f037ee00-63c0-11eb-8364-e38b1a778528.png"  width="200" />

## Edit Profile Screen:
- removed role `Picker`, added `Text` that indicates user's `role`
- added comment to note that this screen requires `uid` passed as a `navigation` parameter
<img src="https://user-images.githubusercontent.com/69874869/106398086-d0ef8f80-63c5-11eb-95d0-c3a763499d25.png"  width="200" />

## Miscellaneous:
- **const.js**:
    - updated `INITIAL_USER_STATE` and `INITIAL_CLASSROOM_STATE` to reflect schema changes. 
    - added `classTypes` and `classDays` enumeration
    - added export for `classTypes` and `classDays` to **const** and **index.js**
- cleaned up some linter issues and added comment to note that this screen requires `uid` passed as a `navigation` parameter in **ProfileScreen.js**
- added comments on **SignInScreen.js**

KarenMTan committed:
Delete .gitignore

KarenMTan committed:
Delete package-lock.json

KarenMTan reviewed (approved):
Looks good! Good work Michael!
